rfd/0156-notifications.md

```diff
@@ -0,0 +1,569 @@
+there may be a quick action button to assume the new roles. Interacting with a notification will mark it as having been clicked on.
+
+##### Stretch goal
+
+Users will also have the ability to manage their notifications preferences in their settings. Particularly, the ability to mute certain types
+of notifications, these will be stored in the user's user preferences. This may not be included in the first version.
+
+#### CLI
+
+##### tctl
+
+Users with RBAC `write` permissions for the `notifications` resource will be able to manually create notifications to notify users in their cluster using:
+
+```shell
+$ tctl notifications create --title=<title> --description=<description> --type=<info | warning> (--all | --roles=<roles-to-target>) --ttl=<ttl>
+```
+
+- `--type` is the type of notification, either `info` or `warning`. All this does is determine which icon/colors will be used when showing the notification in the UI.
+- `--all` means this notification will target all users.
+- `--roles` are the role(s) that this notification should target (ie. `ByRoles`). If multiple roles are specified here, users will only need to have one of the roles to be targeted.
+- `--ttl` is the notification's time to live, it will automatically expire and be deleted after this. The default is 30 days.
+
+###### Example usage
+
+```shell
+$  tctl notifications create \
+   --title="Enroll an MFA device" \
+   --description="We will soon be enforcing MFA in this cluster, please enroll a device to avoid being locked out of your account." \
+   --type=info \
+   --all \
+   --ttl=45d
+```
+
+Global notifications can also be deleted using:
+
+```shell
+$ tctl notifications delete <notification-id>
+```
+
+###### Stretch goal
+
+Support a `--permissions` flag that accepts a predicate expression which will allow cluster admins to also be able to target users by permissions.
+
+### Security
+
+The proposed changes don't introduce any new potential vulnerabilities, though it should be noted that there is already a relevant minor issue affecting all WebSocket connections in the UI, which is
+being tracked by [this ticket](https://github.com/gravitational/teleport-private/issues/637) and will be remediated with the implementation of [this RFD](https://github.com/gravitational/teleport.e/pull/1609).
+
+Creating a notification using `tctl` is a protected action that requires `write` permissions for the `notifications` resource which users won't have by default. Creating notifications will also be
+an [administrative action](https://github.com/gravitational/teleport/blob/master/rfd/0131-adminitrative-actions-mfa.md) and require an additional MFA re-authentication. This prevents an unauthorized bad actor from creating a malicious notification, such as to phish users.
```

Comment:
It's always best to assume that someone _will_ end up creating a malicious notification at some point, and we should do what we can to make their lives difficult.  As we've seen from previous audits of the access request reason field and the cluster alert message field, displaying arbitrary user-provided text is risky.  At the very least we should be disallowing control characters and ensuring that the web UI always correctly escapes all user-determinable fields.  Might also be good to normalize whitespace and have a fairly conservative character limit.